### PR TITLE
Add make test-kitodo

### PR DESCRIPTION
**Pull request by markusweigelt**

### CODE REVIEW

**Comment by markusweigelt** [comment]
Atm we call `make test `and run all test. Possibly only `test-kitodo` and `test-presentation` is still necessary as the `test-production` is indirectly tested by `test-kitodo`.

File: Makefile [hunk 1]
```diff
@@ -144,6 +147,26 @@ config:
...
+	  wget -P $@/images https://digital.slub-dresden.de/data/kitodo/LankDres_1760234508/LankDres_1760234508_tif/jpegs/$$page.tif.original.jpg; \
+	done
+
+test: test-kitodo
+
+test-kitodo: ./kitodo/data/metadata/testdata-kitodo
+# wait until Kitodo.Production directory structure is initialized
+	docker exec -t `docker container ls -qf name=kitodo-app` bash -c "/wait-for-it.sh -t 0 kitodo-app:$$APP_PORT"
+# run asynchronous ocr processing, which should return within 5 seconds with exit status 1
+	timeout --preserve-status 5 docker exec -t `docker container ls -qf name=kitodo-app` bash -c '/usr/local/kitodo/scripts/script_ocr_process_dir.sh "testdata-kitodo" 1'; test $$? -eq 1
```
Comment (bertsky) [inline Makefile:164]: Does this really work? (The POSIX sh's test builtin AFAIK does not know `-eq`. But perhaps this is a dashism?)

```suggestion
	timeout --preserve-status 5 docker exec -t `docker container ls -qf name=kitodo-app` bash -c '/usr/local/kitodo/scripts/script_ocr_process_dir.sh "testdata-kitodo" 1'; test $$? = 1
```

File: Makefile [hunk 2]
```diff
@@ -144,6 +147,26 @@ config:
...
+# Check the docker log of the manager if the ocr processing has already been completed. It fails if the ocr processing was not completed within 5 minutes.
+	( timeout 5m docker logs -f `docker container ls -qf name=ocrd-manager` & ) | grep -q "KitodoActiveMQClient"
```
Comment (bertsky) [inline Makefile:166]: Very nice recipe!

File: Makefile [hunk 3]
```diff
@@ -144,6 +147,26 @@ config:
...
+# Test if the alto files exist
+	test -d ./kitodo/data/metadata/testdata-kitodo/ocr/alto
+	test -s ./kitodo/data/metadata/testdata-kitodo/ocr/alto/00000009.tif.original.xml
```
Comment (bertsky) [inline Makefile:169]:
```suggestion
	test -d $(APP_DATA)/metadata/testdata-kitodo/ocr/alto
	test -s $(APP_DATA)/metadata/testdata-kitodo/ocr/alto/00000009.tif.original.xml
```

**Review by bertsky — COMMENTED** [review]
Excellent!

**Comment by bertsky** [comment]
> Possibly only `test-kitodo` and `test-presentation` is still necessary as the `test-production` is indirectly tested by `test-kitodo`.

Yes, but `test-production` (alone) tests in sync mode, which is good for coverage.

BTW, on the CI we could do `make -j test` which will run all 3 tests in parallel. To make that work with the Controller, we can either add `CONTROLLER_WORKERS=3 make -j test` so that they will actually be computed in parallel, or keep the default 1 (which will have each job wait for the previous job to free the semaphore). We could even test both...

File: Makefile (hunk 1 above)
Comment (markusweigelt) [reply Makefile:164]: Yes it worked. It tested this with exit status 124 of timeout in my local environment.

Comment (bertsky) [reply Makefile]: Ok, then the CI uses dash by default. Better with the suggestion, that should work on both dash/ksh and bash/sh.

**Comment by markusweigelt** [comment]
> BTW, on the CI we could do `make -j test` which will run all 3 tests in parallel. To make that work with the Controller, we can either add `CONTROLLER_WORKERS=3 make -j test` so that they will actually be computed in parallel, or keep the default 1 (which will have each job wait for the previous job to free the semaphore). We could even test both...

Probably we have to adjust the make call in Makefile too https://github.com/slub/ocrd_kitodo/blob/6f665a2c8e550fd5a2498d501aaeb950017ac9c6/Makefile#L153.

Atm rerunning of `test-kitodo` without `make down` and `make start` is a problem cause we need to flush the manager log to provide the check for `KitodoActiveMQClient` if ocr processing has finished. https://github.com/slub/ocrd_kitodo/blob/c3d29a5713aa3894e75117a27da01e3768b92221/Makefile#L169

**Comment by bertsky** [comment]
> Atm rerunning of `test-kitodo` without `make down` and `make start` is a problem cause we need to flush the manager log to provide the check for `KitodoActiveMQClient` if ocr processing has finished.

indeed.

But truncating the Docker log file is asking for trouble. I'd rather keep it simple and require down-up.

Or is there another way to query the ActiveMQ was received, perhaps something on the process directory?

**Comment by markusweigelt** [comment]
Yes. Instead of testing for existing `KitodoActiveMQClient` request in log, we can check if the alto files are already in the directory. If they are not present within of 5 minutes, an error will be thrown. We would only have to delete the alto directory before rerun next test.

**Comment by bertsky** [comment]
Yes, sounds good. (Or directly intercept the ActiveMQ result by interacting with our kitodo-mq container?)

**Comment by markusweigelt** [comment]
Changed the implementation to check for exiting alto file.

**Review by bertsky — COMMENTED** [review]
But I really like the ActiveMQ test better! 

(Even if that means you cannot easily repeat. And why not try to test directly for the ActiveMQ reply? For example, we could do `docker exec kitodo-mq bin/activemq consumer --messageCount 1` ...)

File: Makefile [hunk 4]
```diff
@@ -144,6 +150,28 @@ config:
...
+
+test-kitodo: $(APP_DATA)/metadata/testdata-kitodo
+# remove ocr directory if exist
+	rm -rf $(APP_DATA)/metadata/testdata-kitodo/ocr
+# wait until Kitodo.Production directory structure is initialized
+	docker exec -t `docker container ls -qf name=kitodo-app` bash -c "/wait-for-it.sh -t 0 kitodo-app:$$APP_PORT"
+# run asynchronous ocr processing, which should return within 5 seconds with exit status 1
+	timeout --preserve-status 5 docker exec -t `docker container ls -qf name=kitodo-app` bash -c '/usr/local/kitodo/scripts/script_ocr_process_dir.sh "testdata-kitodo" 1'; test $$? = 1
+# check with interval of 1 second if ocr folder exists. It fails if the ocr folder is not created within 5 minutes.
+	timeout 5m bash -c 'until ls $(APP_DATA)/metadata/testdata-kitodo -R | grep -m 1 "00000014.tif.original.xml"; do : sleep 5; done'
```
[the remark this answers is not in the document]
Comment (markusweigelt) [reply Makefile:171]: > (Even if that means you cannot easily repeat. And why not try to test directly for the ActiveMQ reply? For example, we could do `docker exec kitodo-mq bin/activemq consumer --messageCount 1` ...)

I can do that but don't we run into the same problem as before, without flushing or saving the message count for the rerun?

[the remark this answers is not in the document]
Comment (bertsky) [reply Makefile]: No, I believe this is different than parsing the logs. And if necessary, we could even restart kitodo-mq.

**Comment by markusweigelt** [comment]
Add the issue #64 and merge branch as discussed

**Pull request merged**